File: python/src/gitea_toolkit_mcp/tools/action_tools.py
"""Gitea Actions 域工具（对应 TS 版 `src/ai/tools/actionTools.ts`）。

能力边界（与 Gitea 1.26.4 的 API 一致，不要凭空加）：

- **没有「取消运行」**：只有 ``rerun`` / ``rerun-failed-jobs``。
  ``DELETE /actions/runs/{run}`` 是删除记录，语义不同，这里刻意不暴露。
- 产物**只列不下载**：下载是 zip 二进制流，不适合作为工具返回值。

参数顺序：必填在前且不带默认值（详细理由见 `issue_tools.py` 顶部说明）。
"""

from __future__ import annotations

import json
from typing import Annotated, Any

from pydantic import Field

from ..operations import ActionOperations
from ._format import (
    action_state_label,
    format_artifact_line,
    format_job_line,
    format_run_line,
    format_workflow_line,
    relative_time,
)
#: 起别名是因为下面有个工具参数就叫 tail_lines（参数名即 schema 里的属性名，
#: 必须与 TS 版一致），不能与助手函数撞名
from ._format import tail_lines as _tail_lines
from ._registry import tool
from ._shared import OWNER_ARG, REPO_ARG, ctx, resolve_repo

_owner = OWNER_ARG
_repo = REPO_ARG


def register(server: Any) -> None:
    """注册 Actions 域全部工具。"""

    @tool(server, name="gitea_list_workflows", title="列出工作流")
    async def gitea_list_workflows(owner: _owner = None, repo: _repo = None) -> str:
        """列出仓库的全部工作流（即 .gitea/workflows 下的 YAML），包含启用状态与工作流 ID。"""
        c = ctx()
        target = resolve_repo(owner, repo)
        workflows = await ActionOperations(c.client).list_workflows(
            target.owner, target.repo
        )
        if not workflows:
            return (
                f"{target.full_name} 没有配置任何工作流"
                "（该接口只枚举 `.gitea/workflows/`，不包含 `.github/workflows/`）。"
            )
        lines = "\n".join(format_workflow_line(item) for item in workflows)
        return c.truncate(f"{target.full_name} 共 {len(workflows)} 个工作流：\n{lines}")

    @tool(server, name="gitea_list_action_runs", title="列出工作流运行记录")
    async def gitea_list_action_runs(
        owner: _owner = None,
        repo: _repo = None,
        event: Annotated[
            str | None,
            Field(
                description="触发事件过滤，例如 push / pull_request / workflow_dispatch / schedule。"
            ),
        ] = None,
        branch: Annotated[str | None, Field(description="按分支过滤。")] = None,
        status: Annotated[
            str | None,
            Field(description="按状态过滤，例如 success / failure / running / waiting。"),
        ] = None,
        actor: Annotated[str | None, Field(description="按触发者用户名过滤。")] = None,
        head_sha: Annotated[str | None, Field(description="按提交 SHA 过滤。")] = None,
        limit: Annotated[int | None, Field(ge=1, le=100, description="返回数量上限，默认 20。")] = None,
    ) -> str:
        """列出仓库的工作流运行记录，支持按触发事件、分支、状态、触发者、提交 SHA 过滤，用于排查 CI 是否通过。"""
        c = ctx()
        target = resolve_repo(owner, repo)
        result = await ActionOperations(c.client).list_runs(
            target.owner,
            target.repo,
            event=event,
            branch=branch,
            status=status,
            actor=actor,
            head_sha=head_sha,
            limit=limit or 20,
        )
        if not result.items:
            return "没有匹配的运行记录。"
        lines = "\n".join(format_run_line(item) for item in result.items)
        more = (
            "\n\n_还有更多记录，可调整过滤条件或增大 limit。_"
            if result.page_info.has_next_page
            else ""
        )
        return c.truncate(
            f"{target.full_name} 匹配到 {len(result.items)} 条运行记录：\n{lines}{more}"
        )

    @tool(server, name="gitea_get_action_run", title="获取工作流运行详情")
    async def gitea_get_action_run(
        run_id: Annotated[
            int,
            Field(ge=1, description="运行 ID（列表结果里的 run id，不是 # 后面的编号）。"),
        ],
        owner: _owner = None,
        repo: _repo = None,
    ) -> str:
        """获取单次工作流运行（workflow run）的详情，并列出它包含的全部作业及其步骤状态，用于定位是哪一步失败。"""
        c = ctx()
        target = resolve_repo(owner, repo)
        operations = ActionOperations(c.client)
        run = await operations.get_run(target.owner, target.repo, run_id)
        jobs = await operations.list_run_jobs(target.owner, target.repo, run_id)

        header = (
            f"**#{run.get('run_number') or run.get('id')}** {run.get('display_title') or ''}"
            f" —— {action_state_label(run.get('status'), run.get('conclusion'))}"
        )
        meta = " · ".join(
            part
            for part in (
                f"事件 `{run['event']}`" if run.get("event") else "",
                f"分支 `{run['head_branch']}`" if run.get("head_branch") else "",
                f"提交 `{str(run.get('head_sha') or '')[:7]}`" if run.get("head_sha") else "",
                f"触发者 @{(run.get('actor') or {}).get('login')}"
                if (run.get("actor") or {}).get("login")
                else "",
                f"开始于 {relative_time(run.get('started_at'))}",
            )
            if part
        )
        job_text = (
            "\n".join(format_job_line(job) for job in jobs) if jobs else "_（暂无作业）_"
        )
        failed = [job for job in jobs if job.get("conclusion") == "failure"]
        hint = (
            "\n\n> 失败作业："
            + "、".join(f"`{job.get('name')}`（job id `{job.get('id')}`）" for job in failed)
            + "，可用 `gitea_get_job_logs` 查看日志。"
            if failed
            else ""
        )
        return c.truncate(f"{header}\n\n{meta}\n\n作业：\n{job_text}{hint}")

    @tool(server, name="gitea_get_job_logs", title="获取作业日志")
    async def gitea_get_job_logs(
        job_id: Annotated[
            int, Field(ge=1, description="作业 ID（可从 gitea_get_action_run 的结果取得）。")
        ],
        owner: _owner = None,
        repo: _repo = None,
        tail_lines: Annotated[
            int | None, Field(ge=10, le=2000, description="保留末尾多少行，默认 200。")
        ] = None,
    ) -> str:
        """获取某个工作流作业的原始日志文本。默认只保留**末尾**若干行，因为失败信息通常出现在最后。"""
        c = ctx()
        target = resolve_repo(owner, repo)
        raw = await ActionOperations(c.client).get_job_logs(
            target.owner, target.repo, job_id
        )
        if not raw.strip():
            return f"作业 `{job_id}` 没有日志（可能仍在排队，或日志已被清理）。"
        text, truncated = _tail_lines(raw, tail_lines or 200)
        note = "\n\n_（日志较长，以上仅为末尾部分）_" if truncated else ""
        return c.truncate(f"作业 `{job_id}` 的日志：\n\n```\n{text}\n```{note}")

    @tool(server, name="gitea_list_artifacts", title="列出构建产物")
    async def gitea_list_artifacts(
        owner: _owner = None,
        repo: _repo = None,
        name: Annotated[str | None, Field(description="按产物名精确过滤。")] = None,
        limit: Annotated[int | None, Field(ge=1, le=100, description="返回数量上限，默认 20。")] = None,
    ) -> str:
        """列出仓库的工作流构建产物（artifact），包含大小与是否已过期。"""
        c = ctx()
        target = resolve_repo(owner, repo)
        result = await ActionOperations(c.client).list_artifacts(
            target.owner, target.repo, name=name, limit=limit or 20
        )
        if not result.items:
            return "没有匹配的构建产物。"
        lines = "\n".join(format_artifact_line(item) for item in result.items)
        return c.truncate(
            f"{target.full_name} 匹配到 {len(result.items)} 个产物：\n{lines}"
        )

    @tool(server, name="gitea_dispatch_workflow", title="触发工作流", read_only=False, destructive=False)
    async def gitea_dispatch_workflow(
        workflow_id: Annotated[
            str,
            Field(min_length=1, description="工作流 ID 或文件名，例如 `ci.yml`。可用 gitea_list_workflows 查询。"),
        ],
        ref: Annotated[str, Field(min_length=1, description="目标引用：分支名、标签名或提交 SHA。")],
        owner: _owner = None,
        repo: _repo = None,
        inputs: Annotated[
            dict[str, str] | None,
            Field(description="传给工作流的输入键值对（字符串），对应 workflow_dispatch 声明的 inputs。"),
        ] = None,
    ) -> str:
        """手动触发一个声明了 `on: workflow_dispatch` 的 Gitea 工作流，可指定分支/标签/提交与 inputs。若工作流未声明 workflow_dispatch，服务端会拒绝。"""
        c = ctx()
        target = resolve_repo(owner, repo)
        await ActionOperations(c.client).dispatch_workflow(
            target.owner,
            target.repo,
            workflow_id=workflow_id,
            ref=ref,
            inputs=inputs,
        )
        input_note = ""
        if inputs:
            input_note = (
                "\n\n传入的 inputs：\n```json\n"
                + json.dumps(inputs, ensure_ascii=False, indent=2)
                + "\n```"
            )
        return c.truncate(
            f"已触发工作流 `{workflow_id}`（ref `{ref}`）于 {target.full_name}。"
            f"运行需要一点时间才会出现在列表中。{input_note}"
        )

    @tool(server, name="gitea_rerun_action", title="重新运行工作流", read_only=False, destructive=False)
    async def gitea_rerun_action(
        run_id: Annotated[int, Field(ge=1, description="运行 ID。")],
        owner: _owner = None,
        repo: _repo = None,
        failed_only: Annotated[
            bool | None,
            Field(description="设为 true 时只重跑失败的作业，默认 false（整条重跑）。"),
        ] = None,
    ) -> str:
        """重新运行一次工作流运行记录：可整条重跑，也可只重跑失败的作业。注意 Gitea 的 API **不提供取消运行**，只有重跑。"""
        c = ctx()
        target = resolve_repo(owner, repo)
        failed_only_value = failed_only is True
        operations = ActionOperations(c.client)
        if failed_only_value:
            await operations.rerun_failed_jobs(target.owner, target.repo, run_id)
        else:
            await operations.rerun_run(target.owner, target.repo, run_id)
        return c.truncate(
            f"已提交重跑请求：{target.full_name} 运行 `{run_id}`"
            f"（{'仅失败的作业' if failed_only_value else '整条运行'}）。"
        )

    @tool(server, name="gitea_set_workflow_enabled", title="启用/停用工作流", read_only=False)
    async def gitea_set_workflow_enabled(
        workflow_id: Annotated[
            str, Field(min_length=1, description="工作流 ID 或文件名，例如 `ci.yml`。")
        ],
        enabled: Annotated[bool, Field(description="true 为启用，false 为停用。")],
        owner: _owner = None,
        repo: _repo = None,
    ) -> str:
        """启用或停用一个 Gitea 工作流。停用后该工作流不会响应任何触发事件。"""
        c = ctx()
        target = resolve_repo(owner, repo)
        await ActionOperations(c.client).set_workflow_enabled(
            target.owner, target.repo, workflow_id, enabled
        )
        return c.truncate(
            f"已{'启用' if enabled else '停用'}工作流 `{workflow_id}`（{target.full_name}）。"
        )


__all__ = ["register"]

File: src/ai/tools/actionTools.ts
/**
 * Gitea Actions 域 AI 工具。
 *
 * 能力边界（与 Gitea 1.26.4 的 API 一致，不要凭空加）：
 *   - **没有「取消运行」**：Gitea 只提供 rerun / rerun-failed-jobs。
 *     `DELETE /actions/runs/{run}` 是删除记录，语义不同，这里刻意不暴露。
 *   - 产物只列不下载：下载是 zip 二进制流，不适合作为工具返回值；
 *     需要时用 `archive_download_url` 或走网页。
 */
import { z } from 'zod';
import { defineTool, type GiteaToolDefinition } from './types';
import {
  actionStateLabel,
  formatArtifactLine,
  formatJobLine,
  formatRunLine,
  formatWorkflowLine,
  relativeTime,
  resolveRepoRef,
  tailLines,
} from './helpers';

/** owner / repo 入参的公共片段。 */
const repoArgs = {
  owner: z.string().optional().describe('仓库所属者。省略时使用当前工作区推断出的仓库。'),
  repo: z.string().optional().describe('仓库名。省略时使用当前工作区推断出的仓库。'),
};

/** Actions 域工具集合。 */
export const actionTools: GiteaToolDefinition[] = [
  defineTool({
    name: 'gitea_list_workflows',
    displayName: '列出工作流',
    description:
      '列出仓库的全部工作流（即 .gitea/workflows 下的 YAML），包含启用状态与工作流 ID。',
    userDescription: '列出 Gitea 工作流',
    category: 'action',
    access: 'read',
    inputShape: { ...repoArgs },
    handler: async (input, ctx) => {
      const { owner, repo } = resolveRepoRef(input, ctx);
      const workflows = await ctx.operations.actions.listWorkflows({ owner, repo });
      if (workflows.length === 0) {
        return {
          text: `${owner}/${repo} 没有配置任何工作流（应在 \`.gitea/workflows/\` 下）。`,
          data: { items: [] },
        };
      }
      return {
        text: `${owner}/${repo} 共 ${workflows.length} 个工作流：\n${workflows.map(formatWorkflowLine).join('\n')}`,
        data: { items: workflows },
      };
    },
  }),

  defineTool({
    name: 'gitea_list_action_runs',
    displayName: '列出工作流运行记录',
    description:
      '列出仓库的工作流运行记录，支持按触发事件、分支、状态、触发者、提交 SHA 过滤，用于排查 CI 是否通过。',
    userDescription: '列出工作流运行记录',
    category: 'action',
    access: 'read',
    inputShape: {
      ...repoArgs,
      event: z
        .string()
        .optional()
        .describe('触发事件过滤，例如 push / pull_request / workflow_dispatch / schedule。'),
      branch: z.string().optional().describe('按分支过滤。'),
      status: z
        .string()
        .optional()
        .describe('按状态过滤，例如 success / failure / running / waiting。'),
      actor: z.string().optional().describe('按触发者用户名过滤。'),
      head_sha: z.string().optional().describe('按提交 SHA 过滤。'),
      limit: z.number().int().min(1).max(100).optional().describe('返回数量上限，默认 20。'),
    },
    handler: async (input, ctx) => {
      const { owner, repo } = resolveRepoRef(input, ctx);
      const result = await ctx.operations.actions.listRuns({
        owner,
        repo,
        event: input.event,
        branch: input.branch,
        status: input.status,
        actor: input.actor,
        headSha: input.head_sha,
        limit: input.limit ?? 20,
      });
      if (result.items.length === 0) {
        return { text: '没有匹配的运行记录。', data: { items: [] } };
      }
      const more = result.pageInfo.hasNextPage ? '\n\n_还有更多记录，可调整过滤条件或增大 limit。_' : '';
      return {
        text: `${owner}/${repo} 匹配到 ${result.items.length} 条运行记录：\n${result.items.map(formatRunLine).join('\n')}${more}`,
        data: { items: result.items, pageInfo: result.pageInfo },
      };
    },
  }),

  defineTool({
    name: 'gitea_get_action_run',
    displayName: '获取工作流运行详情',
    description:
      '获取单次工作流运行（workflow run）的详情，并列出它包含的全部作业及其步骤状态，用于定位是哪一步失败。',
    userDescription: '查看工作流运行详情',
    category: 'action',
    access: 'read',
    inputShape: {
      ...repoArgs,
      run_id: z.number().int().positive().describe('运行 ID（列表结果里的 run id，不是 # 后面的编号）。'),
    },
    handler: async (input, ctx) => {
      const { owner, repo } = resolveRepoRef(input, ctx);
      const [run, jobs] = await Promise.all([
        ctx.operations.actions.getRun(owner, repo, input.run_id),
        ctx.operations.actions.listRunJobs(owner, repo, input.run_id),
      ]);
      const header = `**#${run.run_number ?? run.id}** ${run.display_title ?? ''} —— ${actionStateLabel(run.status, run.conclusion)}`;
      const meta = [
        run.event ? `事件 \`${run.event}\`` : '',
        run.head_branch ? `分支 \`${run.head_branch}\`` : '',
        run.head_sha ? `提交 \`${run.head_sha.slice(0, 7)}\`` : '',
        run.actor?.login ? `触发者 @${run.actor.login}` : '',
        `开始于 ${relativeTime(run.started_at)}`,
      ]
        .filter(Boolean)
        .join(' · ');
      const failed = jobs.filter((job) => job.conclusion === 'failure');
      const jobText =
        jobs.length === 0 ? '_（暂无作业）_' : jobs.map(formatJobLine).join('\n');
      const hint =
        failed.length > 0
          ? `\n\n> 失败作业：${failed.map((job) => `\`${job.name}\`（job id \`${job.id}\`）`).join('、')}，可用 \`gitea_get_job_logs\` 查看日志。`
          : '';
      return {
        text: `${header}\n\n${meta}\n\n作业：\n${jobText}${hint}`,
        data: { run, jobs },
      };
    },
  }),

  defineTool({
    name: 'gitea_get_job_logs',
    displayName: '获取作业日志',
    description:
      '获取某个工作流作业的原始日志文本。默认只保留**末尾**若干行，因为失败信息通常出现在最后。',
    userDescription: '查看作业日志',
    category: 'action',
    access: 'read',
    inputShape: {
      ...repoArgs,
      job_id: z.number().int().positive().describe('作业 ID（可从 gitea_get_action_run 的结果取得）。'),
      tail_lines: z
        .number()
        .int()
        .min(10)
        .max(2000)
        .optional()
        .describe('保留末尾多少行，默认 200。'),
    },
    handler: async (input, ctx) => {
      const { owner, repo } = resolveRepoRef(input, ctx);
      const raw = await ctx.operations.actions.getJobLogs(owner, repo, input.job_id);
      if (raw.trim().length === 0) {
        return { text: `作业 \`${input.job_id}\` 没有日志（可能仍在排队，或日志已被清理）。`, data: {} };
      }
      const { text, truncated } = tailLines(raw, input.tail_lines ?? 200);
      const note = truncated ? '\n\n_（日志较长，以上仅为末尾部分）_' : '';
      return {
        text: `作业 \`${input.job_id}\` 的日志：\n\n\`\`\`\n${text}\n\`\`\`${note}`,
        data: { jobId: input.job_id, truncated },
      };
    },
  }),

  defineTool({
    name: 'gitea_list_artifacts',
    displayName: '列出构建产物',
    description: '列出仓库的工作流构建产物（artifact），包含大小与是否已过期。',
    userDescription: '列出构建产物',
    category: 'action',
    access: 'read',
    inputShape: {
      ...repoArgs,
      name: z.string().optional().describe('按产物名精确过滤。'),
      limit: z.number().int().min(1).max(100).optional().describe('返回数量上限，默认 20。'),
    },
    handler: async (input, ctx) => {
      const { owner, repo } = resolveRepoRef(input, ctx);
      const result = await ctx.operations.actions.listArtifacts({
        owner,
        repo,
        name: input.name,
        limit: input.limit ?? 20,
      });
      if (result.items.length === 0) {
        return { text: '没有匹配的构建产物。', data: { items: [] } };
      }
      return {
        text: `${owner}/${repo} 匹配到 ${result.items.length} 个产物：\n${result.items.map(formatArtifactLine).join('\n')}`,
        data: { items: result.items, pageInfo: result.pageInfo },
      };
    },
  }),

  defineTool({
    name: 'gitea_dispatch_workflow',
    displayName: '触发工作流',
    description:
      '手动触发一个声明了 `on: workflow_dispatch` 的 Gitea 工作流，可指定分支/标签/提交与 inputs。' +
      '若工作流未声明 workflow_dispatch，服务端会拒绝。',
    userDescription: '手动触发 Gitea 工作流',
    category: 'action',
    access: 'write',
    inputShape: {
      ...repoArgs,
      workflow_id: z
        .string()
        .describe('工作流 ID 或文件名，例如 `ci.yml`。可用 gitea_list_workflows 查询。'),
      ref: z.string().describe('目标引用：分支名、标签名或提交 SHA。'),
      inputs: z
        .record(z.string())
        .optional()
        .describe('传给工作流的输入键值对（字符串），对应 workflow_dispatch 声明的 inputs。'),
    },
    handler: async (input, ctx) => {
      const { owner, repo } = resolveRepoRef(input, ctx);
      await ctx.operations.actions.dispatchWorkflow({
        owner,
        repo,
        workflowId: input.workflow_id,
        ref: input.ref,
        inputs: input.inputs,
      });
      const inputNote =
        input.inputs && Object.keys(input.inputs).length > 0
          ? `\n\n传入的 inputs：\n\`\`\`json\n${JSON.stringify(input.inputs, null, 2)}\n\`\`\``
          : '';
      return {
        text: `已触发工作流 \`${input.workflow_id}\`（ref \`${input.ref}\`）于 ${owner}/${repo}。运行需要一点时间才会出现在列表中。${inputNote}`,
        data: { workflowId: input.workflow_id, ref: input.ref },
      };
    },
  }),

  defineTool({
    name: 'gitea_rerun_action',
    displayName: '重新运行工作流',
    description:
      '重新运行一次工作流运行记录：可整条重跑，也可只重跑失败的作业。' +
      '注意 Gitea 的 API **不提供取消运行**，只有重跑。',
    userDescription: '重新运行 Gitea 工作流',
    category: 'action',
    access: 'write',
    inputShape: {
      ...repoArgs,
      run_id: z.number().int().positive().describe('运行 ID。'),
      failed_only: z
        .boolean()
        .optional()
        .describe('设为 true 时只重跑失败的作业，默认 false（整条重跑）。'),
    },
    handler: async (input, ctx) => {
      const { owner, repo } = resolveRepoRef(input, ctx);
      const failedOnly = input.failed_only === true;
      if (failedOnly) {
        await ctx.operations.actions.rerunFailedJobs(owner, repo, input.run_id);
      } else {
        await ctx.operations.actions.rerunRun(owner, repo, input.run_id);
      }
      return {
        text: `已提交重跑请求：${owner}/${repo} 运行 \`${input.run_id}\`（${failedOnly ? '仅失败的作业' : '整条运行'}）。`,
        data: { runId: input.run_id, failedOnly },
      };
    },
  }),

  defineTool({
    name: 'gitea_set_workflow_enabled',
    displayName: '启用/停用工作流',
    description: '启用或停用一个 Gitea 工作流。停用后该工作流不会响应任何触发事件。',
    userDescription: '启用或停用 Gitea 工作流',
    category: 'action',
    access: 'write',
    inputShape: {
      ...repoArgs,
      workflow_id: z.string().describe('工作流 ID 或文件名，例如 `ci.yml`。'),
      enabled: z.boolean().describe('true 为启用，false 为停用。'),
    },
    handler: async (input, ctx) => {
      const { owner, repo } = resolveRepoRef(input, ctx);
      await ctx.operations.actions.setWorkflowEnabled(owner, repo, input.workflow_id, input.enabled);
      return {
        text: `已${input.enabled ? '启用' : '停用'}工作流 \`${input.workflow_id}\`（${owner}/${repo}）。`,
        data: { workflowId: input.workflow_id, enabled: input.enabled },
      };
    },
  }),
];
